### PR TITLE
Quick (& dirty) POC of saving memory with GBM trees

### PR DESCRIPTION
**Pull request by michalkurka**
This PR introduces the idea of saving for Keyed objects that are essentially wrappers around a byte-array.

The perfect example of such an object is CompressedTree which represents both the scoring and auxiliary tree. By using a LightKeyed object we able to re-use the bytes stored in Value object and thus reducing the needed memory to a half.

### CODE REVIEW

File: h2o-core/src/main/java/water/MemKeyed.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package water;
+
+public class MemKeyed<T extends MemKeyed<T>> extends Keyed<T> {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/MemKeyed.java:3]: So this is something like `Chunk` if i understand well?

File: h2o-core/src/main/java/water/MemKeyed.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  public MemKeyed(Key<T> key, byte[] mem) {
+    super(key);
+    _mem = mem;
+  }
+
+  @SuppressWarnings("unchecked")
+  public final T reloadFromBytes(Key<T> k, byte[] ary) {
+    _key = k;
+    _mem = ary;
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/MemKeyed.java:15]: Just to make sure i understand correctly, this is the key trick - simply do not materialize a new object from `_mem` but simply just keep reference (which is shared from `Value`).

File: h2o-core/src/main/java/water/MemKeyed.java (hunk 1 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/MemKeyed.java:3]: this is the same pattern as chunk except for it also is Keyed which Chunk itself isn't

File: h2o-core/src/main/java/water/MemKeyed.java (hunk 2 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/MemKeyed.java:15]: yes, precisely - that is the trick

**Pull request closed without merging**